mono/metadata/threadpool-netcore.c

```diff
@@ -0,0 +1,625 @@
+			counter._.working --;
+		});
+
+		mono_refcount_dec (&threadpool);
+		return;
+	}
+
+	/*
+	 * This is needed so there is always an lmf frame in the runtime invoke call below,
+	 * so ThreadAbortExceptions are caught even if the thread is in native code.
+	 */
+	mono_defaults.threadpool_perform_wait_callback_method->save_lmf = TRUE;
+
+	tp_lock ();
+
+	gsize name_generation = thread->name.generation;
+	/* Set the name if this is the first call to worker_callback on this thread */
+	if (name_generation == 0)
+	   name_generation = set_thread_name (thread);
+
+	while (!mono_runtime_is_shutting_down ()) {
+		gboolean retire = FALSE;
+
+		if (thread->state & (ThreadState_AbortRequested | ThreadState_SuspendRequested)) {
+			tp_unlock ();
+			if (mono_thread_interruption_checkpoint_bool ()) {
+				tp_lock ();
+				continue;
+			}
+			tp_lock ();
+		}
+
+		if (threadpool.outstanding_request == 0)
+			break;
+
+		threadpool.outstanding_request --;
+		g_assert (threadpool.outstanding_request >= 0);
+
+		mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_THREADPOOL, "[%p] worker (outstanding requests %d)",
+			GUINT_TO_POINTER (MONO_NATIVE_THREAD_ID_TO_UINT (mono_native_thread_id_get ())), threadpool.outstanding_request);
+
+		g_assert (threadpool.threadpool_jobs >= 0);
+		threadpool.threadpool_jobs ++;
+
+		tp_unlock ();
+
+		// Any thread can set any other thread name at any time.
+		// So this is unavoidably racy.
+		// This only partly fights against that -- i.e. not atomic and not a loop.
+		// It is reliable against the thread setting its own name, and somewhat
+		// reliable against other threads setting this thread's name.
+		if (name_generation != thread->name.generation)
+			name_generation = set_thread_name (thread);
+
+		mono_thread_clear_and_set_state (thread,
+			(MonoThreadState)~ThreadState_Background,
+			ThreadState_Background);
+
+		MonoObject *exc = NULL, *res;
+
+		res = try_invoke_perform_wait_callback (&exc, error);
+		if (exc || !is_ok (error)) {
+			if (exc == NULL)
+				exc = (MonoObject *) mono_error_convert_to_exception (error);
+			else
+				mono_error_cleanup (error);
+			mono_thread_internal_unhandled_exception (exc);
+		} else if (res && *(MonoBoolean*) mono_object_unbox_internal (res) == FALSE) {
+			retire = TRUE;
+		}
+
+		/* Reset name after every callback */
+		if (name_generation != thread->name.generation)
+			name_generation = set_thread_name (thread);
+
+		tp_lock ();
+
+		threadpool.threadpool_jobs --;
+		g_assert (threadpool.threadpool_jobs >= 0);
+
+		if (retire)
+			break;
+	}
+
+	tp_unlock ();
+
+	COUNTER_ATOMIC (counter, {
+		counter._.working --;
+	});
+
+	mono_refcount_dec (&threadpool);
+}
+
+void
+mono_threadpool_cleanup (void)
+{
+#ifndef DISABLE_SOCKETS
+	mono_threadpool_io_cleanup ();
+#endif
+	mono_lazy_cleanup (&status, cleanup);
+}
+
+MonoAsyncResult *
+mono_threadpool_begin_invoke (MonoDomain *domain, MonoObject *target, MonoMethod *method, gpointer *params, MonoError *error)
```

Comment:
Is this function also going to always be called with the root domain, like the above one? If so, probably add a similar assert.

Reply:
Or maybe not so that you aren't forced to go through another round of CI...